File: Java/MergeSort.java
import java.util.Scanner;

public class MergeSort {
    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter the number of elements : ");
        int n = sc.nextInt();
        int arr[] = new int[n];
        System.out.println("Enter " + n + " elements :");
        for (int i = 0; i < n; i++)
            arr[i] = sc.nextInt();

        mergeSort(arr, 0, n - 1);
        System.out.println("\nThe sorted array : ;");
        for (int i = 0; i < n; i++)
            System.out.print(arr[i] + " ");
        System.out.println();
    }

    static void mergeSort(int arr[], int lower, int upper) {
        if (lower >= upper)
            return;
        int m = (lower + upper) / 2;
        mergeSort(arr, lower, m);
        mergeSort(arr, m + 1, upper);
        merge(arr, lower, upper);
    }

    private static void merge(int arr[], int lower, int upper) {
        int m = (lower + upper) / 2;
        int a[] = new int[m - lower + 1];
        int b[] = new int[upper - m];
        int i, k = 0, k1 = 0, k2 = 0;
        for (i = lower; i <= m; i++, k++)
            a[k] = arr[i];
        k = 0;
        for (; i <= upper; i++, k++)
            b[k] = arr[i];
        for (i = lower; i <= upper && k1 < m - lower + 1 && k2 < upper - m; i++) {
            if (a[k1] < b[k2]) {
                arr[i] = a[k1];
                k1++;
            } else {
                arr[i] = b[k2];
                k2++;
            }
        }
        for (; k1 < m - lower + 1; k1++)
            arr[i++] = a[k1];
        for (; k2 < upper - m; k2++)
            arr[i++] = b[k2];
    }
}

File: Kotlin/MergeSort.kt
import java.util.Scanner

class MergeSort {
    fun main(args: Array<String>) {
        val sc = Scanner(System.`in`)
        System.out.print("Enter the number of elements : ")
        val n = sc.nextInt()
        val arr = IntArray(n)
        System.out.println("Enter $n elements :")
        for (i in 0 until n)
            arr[i] = sc.nextInt()

        mergeSort(arr, 0, n - 1)
        System.out.println("\nThe sorted array : ;")
        for (i in 0 until n)
            System.out.print(arr[i].toString() + " ")
        System.out.println()
    }

    internal fun mergeSort(arr: IntArray, lower: Int, upper: Int) {
        if (lower >= upper)
            return
        val m = (lower + upper) / 2
        mergeSort(arr, lower, m)
        mergeSort(arr, m + 1, upper)
        merge(arr, lower, upper)
    }

    private fun merge(arr: IntArray, lower: Int, upper: Int) {
        val m = (lower + upper) / 2
        val a = IntArray(m - lower + 1)
        val b = IntArray(upper - m)
        var i: Int
        var k = 0
        var k1 = 0
        var k2 = 0
        i = lower
        while (i <= m) {
            a[k] = arr[i]
            i++
            k++
        }
        k = 0
        while (i <= upper) {
            b[k] = arr[i]
            i++
            k++
        }
        i = lower
        while (i <= upper && k1 < m - lower + 1 && k2 < upper - m) {
            if (a[k1] < b[k2]) {
                arr[i] = a[k1]
                k1++
            } else {
                arr[i] = b[k2]
                k2++
            }
            i++
        }
        while (k1 < m - lower + 1) {
            arr[i++] = a[k1]
            k1++
        }
        while (k2 < upper - m) {
            arr[i++] = b[k2]
            k2++
        }
    }
}
